File: modules/shared/dart/cipher_service.dart
// ChaCha20-Poly1305-IETF Cipher Integration Service (Dart)
import 'dart:convert';
import 'dart:typed_data';
import 'package:flutter_sodium/flutter_sodium.dart';

class CipherService {
  static final RegExp _idRegex = RegExp(r'^[a-zA-Z0-9]+$');

  /// Encrypts plaintext using ChaCha20-Poly1305-IETF.
  /// AAD: "${conversationId}:${entityId}".
  /// Layout: Base64url(nonce[12] || ciphertext || tag[16]).
  static String encrypt({
    required Uint8List key,
    required String plaintext,
    required String conversationId,
    required String entityId,
  }) {
    // Assert strict alphanumeric IDs to prevent delimiter injection attacks
    if (!_idRegex.hasMatch(conversationId) || !_idRegex.hasMatch(entityId)) {
      throw ArgumentError('CRYPTO_008: Non-alphanumeric conversationId or entityId');
    }

    final String aad = '$conversationId:$entityId';
    final Uint8List nonce = Sodium.randombytesBuf(12);

    final Uint8List ct = ChaCha20Poly1305Ietf.encryptString(
      plaintext,
      nonce,
      key,
      additionalData: aad,
    );

    final Uint8List blob = Uint8List(12 + ct.length)
      ..setRange(0, 12, nonce)
      ..setRange(12, 12 + ct.length, ct);

    return base64Url.encode(blob).replaceAll('=', '');
  }

  /// Decrypts and authenticates a Base64url blob. Returns null on validation or integrity failure.
  static String? decrypt({
    required Uint8List key,
    required String blob,
    required String conversationId,
    required String entityId,
  }) {
    // Assert strict alphanumeric IDs
    if (!_idRegex.hasMatch(conversationId) || !_idRegex.hasMatch(entityId)) {
      return null;
    }

    try {
      // Add padding back if it was stripped
      final String paddedBlob = blob.padRight(blob.length + (4 - blob.length % 4) % 4, '=');
      final Uint8List bytes = base64Url.decode(paddedBlob);
      
      // Minimum length must be 28 bytes (12-byte nonce + 16-byte minimum auth tag/ciphertext)
      if (bytes.length < 28) {
        return null;
      }

      final Uint8List nonce = bytes.sublist(0, 12);
      final Uint8List ct = bytes.sublist(12);
      final String aad = '$conversationId:$entityId';

      return ChaCha20Poly1305Ietf.decryptString(
        ct,
        nonce,
        key,
        additionalData: aad,
      );
    } catch (_) {
      return null; // Gracefully return null on decryption/verification failure
    }
  }
}

File: modules/shared/node/cipher_service.js
// ChaCha20-Poly1305-IETF Cipher Integration Service (Node.js)
const sodium = require('libsodium-wrappers-sumo');

const idRegex = /^[a-zA-Z0-9]+$/;

/**
 * Encrypts plaintext using ChaCha20-Poly1305-IETF.
 * AAD: "${conversationId}:${entityId}".
 * Layout: Base64url(nonce[12] || ciphertext || tag[16]).
 * @param {Buffer|Uint8Array} key - 32 bytes
 * @param {string} plaintext
 * @param {string} conversationId
 * @param {string} entityId
 * @returns {Promise<string>} Base64url-encoded wire blob
 */
async function encrypt(key, plaintext, conversationId, entityId) {
  await sodium.ready;
  
  // Assert strict alphanumeric IDs to prevent delimiter injection attacks
  if (!idRegex.test(conversationId) || !idRegex.test(entityId)) {
    throw new Error('CRYPTO_008: Non-alphanumeric conversationId or entityId');
  }

  const aad = `${conversationId}:${entityId}`;
  const nonce = sodium.randombytes_buf(sodium.crypto_aead_chacha20poly1305_ietf_NPUBBYTES);
  
  const ct = sodium.crypto_aead_chacha20poly1305_ietf_encrypt(
    plaintext,
    aad,
    null,
    nonce,
    key
  );

  const blob = Buffer.concat([Buffer.from(nonce), Buffer.from(ct)]);
  return blob.toString('base64url');
}

/**
 * Decrypts and authenticates a Base64url wire blob. Returns null on validation or integrity failure.
 * @param {Buffer|Uint8Array} key - 32 bytes
 * @param {string} blob - Base64url-encoded
 * @param {string} conversationId
 * @param {string} entityId
 * @returns {Promise<string|null>}
 */
async function decrypt(key, blob, conversationId, entityId) {
  await sodium.ready;

  // Assert strict alphanumeric IDs
  if (!idRegex.test(conversationId) || !idRegex.test(entityId)) {
    return null;
  }

  try {
    const bytes = Buffer.from(blob, 'base64url');
    
    // Minimum length must be 28 bytes (12-byte nonce + 16-byte minimum auth tag/ciphertext)
    if (bytes.length < 28) {
      return null;
    }

    const nonce = bytes.subarray(0, 12);
    const ct = bytes.subarray(12);
    const aad = `${conversationId}:${entityId}`;

    const pt = sodium.crypto_aead_chacha20poly1305_ietf_decrypt(
      null,
      ct,
      aad,
      nonce,
      key
    );

    return Buffer.from(pt).toString('utf8');
  } catch (error) {
    return null; // Gracefully return null on decryption/verification failure
  }
}

module.exports = {
  encrypt,
  decrypt,
};
